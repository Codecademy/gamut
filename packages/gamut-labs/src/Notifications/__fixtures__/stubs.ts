import { Notification } from '@codecademy/gamut';

export const createStubNotification = (
  overrides?: Partial<Notification>
): Notification => ({
  campaign: 'existence_is_pain',
  date: '5 hours',
  id: '1',
  link: 'https://codecademy.com',
  text: "Look at me! I'm Mr. Notification!",
  ...overrides,
});
