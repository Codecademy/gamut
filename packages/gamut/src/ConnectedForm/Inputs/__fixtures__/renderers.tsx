import React from 'react';

import { FormValues } from '../../../Form/types';
import { FormWrapper, FormWrapperProps } from '../..';

type ConnectedFormInputProps = Pick<
  FormWrapperProps<FormValues>,
  'validation' | 'onSubmit'
> & {
  Field: React.ReactNode;
};

export const ConnectedFormInput: React.FC<ConnectedFormInputProps> = ({
  Field,
  validation = 'onSubmit',
  onSubmit,
  ...rest
}) => {
  return (
    <FormWrapper validation={validation} onSubmit={onSubmit} {...rest}>
      {Field}
    </FormWrapper>
  );
};
