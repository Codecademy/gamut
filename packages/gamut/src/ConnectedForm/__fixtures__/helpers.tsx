import React from 'react';

import {
  ConnectedCheckbox,
  ConnectedFormGroup,
  ConnectedInput,
  ConnectedRadioGroupInput,
  ConnectedSelect,
  SubmitButton,
} from '..';

export const PlainConnectedFields = () => {
  return (
    <>
      <ConnectedFormGroup
        name="checkbox"
        label="cool-checkbox"
        field={{
          component: ConnectedCheckbox,
          label: 'cool-checkbox',
        }}
      />
      <SubmitButton variant="secondary" m={32}>
        submit this form.
      </SubmitButton>
      <ConnectedFormGroup
        name="select"
        label="cool-select"
        field={{
          component: ConnectedSelect,
          options: ['one', 'two', 'zero'],
        }}
      />
      <ConnectedFormGroup
        name="input"
        label="cool-input"
        field={{
          component: ConnectedInput,
        }}
      />
      <ConnectedFormGroup
        name="radiogroup"
        label="cool-radiogroup"
        field={{
          component: ConnectedRadioGroupInput,
          options: [
            { label: 'one', value: 'one' },
            { label: 'two', value: 'two' },
            { label: 'three', value: 'three' },
          ],
        }}
      />
    </>
  );
};

export const ValidatedConnectedFields = () => {
  return (
    <>
      <ConnectedFormGroup
        name="checkbox"
        label="cool-checkbox"
        field={{
          component: ConnectedCheckbox,
          label: 'cool-checkbox',
          validation: { required: 'you must click.' },
        }}
      />
      <SubmitButton variant="secondary" m={32}>
        submit this form.
      </SubmitButton>
      <ConnectedFormGroup
        name="select"
        label="cool-select"
        field={{
          component: ConnectedSelect,
          options: ['one', 'two', 'zero'],
          validation: { required: 'you must select.' },
        }}
      />
    </>
  );
};
