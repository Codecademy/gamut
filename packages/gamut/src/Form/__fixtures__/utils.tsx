import {
  CalendarIcon,
  DataTransferVerticalIcon,
  EarthIcon,
} from '@codecademy/gamut-icons';
import { fireEvent } from '@testing-library/dom';
import { act } from 'react';

export const DOWN_ARROW = { keyCode: 40 };

export const selectOptions = ['red', 'yellow', 'green'];

export const selectOptionsObject = {
  red: 'red',
  yellow: 'yellow',
  green: 'green',
  blue: 'blue',
  teal: 'teal',
  orange: 'orange',
};

export const optionsIconsArray = [
  {
    label: 'Data Transfer Vertical Icon',
    value: 'one',
    icon: DataTransferVerticalIcon,
  },
  { label: 'Calendar Icon', value: 'two', icon: CalendarIcon },
  { label: 'Earth Icon', value: 'three', icon: EarthIcon },
];

export const groupedOptions = [
  {
    label: 'Fruits',
    options: [
      { label: 'Apple', value: 'apple' },
      { label: 'Banana', value: 'banana' },
      { label: 'Orange', value: 'orange' },
    ],
  },
  {
    label: 'Vegetables',
    options: [
      { label: 'Carrot', value: 'carrot' },
      { label: 'Broccoli', value: 'broccoli' },
      { label: 'Spinach', value: 'spinach' },
    ],
  },
];

export const groupedOptionsWithIcons = [
  {
    label: 'Data Icons',
    options: [
      {
        label: 'Data Transfer Vertical Icon',
        value: 'data-transfer',
        icon: DataTransferVerticalIcon,
      },
      {
        label: 'Calendar Icon',
        value: 'calendar',
        icon: CalendarIcon,
      },
    ],
  },
  {
    label: 'Navigation Icons',
    options: [
      {
        label: 'Earth Icon',
        value: 'earth',
        icon: EarthIcon,
      },
    ],
  },
];

export const groupedOptionsWithTitles = [
  {
    label: 'Fruits',
    options: [
      { label: 'Apple', value: 'apple', subtitle: 'Red and crunchy' },
      { label: 'Banana', value: 'banana', subtitle: 'Yellow and sweet' },
      { label: 'Orange', value: 'orange', subtitle: 'Citrus and juicy' },
    ],
  },
  {
    label: 'Vegetables',
    options: [
      { label: 'Carrot', value: 'carrot', subtitle: 'Orange and healthy' },
      {
        label: 'Broccoli',
        value: 'broccoli',
        subtitle: 'Green and nutritious',
      },
      { label: 'Spinach', value: 'spinach', subtitle: 'Leafy and iron-rich' },
    ],
  },
];

export const optionsWithAbbreviations = [
  {
    label: 'United States of America',
    value: 'usa',
    abbreviation: 'USA',
  },
  {
    label: 'United Kingdom',
    value: 'uk',
    abbreviation: 'UK',
  },
  {
    label: 'Canada',
    value: 'canada',
    abbreviation: 'CA',
  },
  {
    label: 'Australia',
    value: 'australia',
    abbreviation: 'AU',
  },
];

export const openDropdown = async (view: any) => {
  await act(() => {
    fireEvent.keyDown(view.getByRole('combobox'), DOWN_ARROW);
    return Promise.resolve();
  });
};
