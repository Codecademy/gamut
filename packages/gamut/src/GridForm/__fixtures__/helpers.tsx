import React from 'react';
import { FormProvider, useForm } from 'react-hook-form';

type FormContextProps = { mode: 'onChange' | 'onSubmit' };

export const FormContext: React.FC<FormContextProps> = ({ mode, children }) => {
  const methods = useForm({
    mode,
  });
  return <FormProvider {...methods}>{children}</FormProvider>;
};
