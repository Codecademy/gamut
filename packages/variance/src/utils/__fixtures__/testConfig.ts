export const testConfig = {
  textColor: { property: 'color' },
  bg: { property: 'backgroundColor' },
  borderColor: { property: 'borderColor' },
  borderColorX: {
    property: 'borderColor',
    properties: ['borderLeftColor', 'borderRightColor'],
    scale: 'colors',
  },
  borderColorY: {
    property: 'borderColor',
    properties: ['borderTopColor', 'borderBottomColor'],
    scale: 'colors',
  },
  borderColorLeft: { property: 'borderLeftColor' },
  borderColorRight: { property: 'borderRightColor' },
  borderColorTop: { property: 'borderTopColor' },
  borderColorBottom: { property: 'borderBottomColor' },

  border: { property: 'border' },
  borderX: { property: 'border', properties: ['borderLeft', 'borderRight'] },
  borderY: { property: 'border', properties: ['borderTop', 'borderBottom'] },
  borderTop: { property: 'borderTop' },
  borderRight: { property: 'borderRight' },
  borderBottom: { property: 'borderBottom' },
  borderLeft: { property: 'borderLeft' },
  borderWidth: { property: 'borderWidth' },
  borderWidthX: {
    property: 'borderWidth',
    properties: ['borderLeftWidth', 'borderRightWidth'],
  },
  borderWidthY: {
    property: 'borderWidth',
    properties: ['borderTopWidth', 'borderBottomWidth'],
  },

  borderRadius: { property: 'borderRadius' },
  borderRadiusLeft: {
    property: 'borderRadius',
    properties: ['borderTopLeftRadius', 'borderBottomLeftRadius'],
  },
  borderRadiusTop: {
    property: 'borderRadius',
    properties: ['borderTopLeftRadius', 'borderTopRightRadius'],
  },
  borderRadiusBottom: {
    property: 'borderRadius',
    properties: ['borderBottomLeftRadius', 'borderBottomRightRadius'],
  },
  borderRadiusRight: {
    property: 'borderRadius',
    properties: ['borderTopRightRadius', 'borderBottomRightRadius'],
  },

  borderStyle: { property: 'borderStyle' },
  borderStyleX: {
    property: 'borderStyle',
    properties: ['borderLeftStyle', 'borderRightStyle'],
  },
  borderStyleY: {
    property: 'borderStyle',
    properties: ['borderTopStyle', 'borderBottomStyle'],
  },

  flex: { property: 'flex' },
  flexBasis: { property: 'flexBasis' },

  p: { property: 'padding' },
  px: {
    property: 'padding',
    properties: ['paddingLeft', 'paddingRight'],
    scale: 'spacing',
  },
  py: {
    property: 'padding',
    properties: ['paddingTop', 'paddingBottom'],
    scale: 'spacing',
  },
  pt: { property: 'paddingTop' },
  pb: { property: 'paddingBottom' },
  pr: { property: 'paddingRight' },
  pl: { property: 'paddingLeft' },
  m: { property: 'margin' },
  mx: {
    property: 'margin',
    properties: ['marginLeft', 'marginRight'],
    scale: 'spacing',
  },
  my: {
    property: 'margin',
    properties: ['marginTop', 'marginBottom'],
    scale: 'spacing',
  },
  mt: { property: 'marginTop' },
  mb: { property: 'marginBottom' },
  mr: { property: 'marginRight' },
  ml: { property: 'marginLeft' },
} as const;
