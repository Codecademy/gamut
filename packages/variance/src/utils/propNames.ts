import { BaseProperty } from '../types/config';

const SHORTHAND_PROPERTIES = [
  'border',
  'borderTop',
  'borderBottom',
  'borderLeft',
  'borderRight',
  'borderWidth',
  'borderStyle',
  'borderColor',
  'background',
  'flex',
  'margin',
  'padding',
  'transition',
  'gap',
  'grid',
  'gridArea',
  'gridColumn',
  'gridRow',
  'gridTemplate',
  'overflow',
  'transition',
];

const SORT = {
  A_BEFORE_B: -1,
  B_BEFORE_A: 1,
  EQUAL: 1,
};

const compare = (a: number, b: number) => {
  switch (Math.sign(a - b)) {
    case -1:
      return SORT.B_BEFORE_A;
    case 1:
      return SORT.A_BEFORE_B;
    default:
      return SORT.EQUAL;
  }
};

/**
 * Orders all properties by the most dependent props
 * @param config
 */
export const orderPropNames = (config: Record<string, BaseProperty>) =>
  Object.keys(config).sort((a, b) => {
    const aProp = config?.[a].property;
    const bProp = config?.[b].property;

    const aProps = config?.[a].properties?.length ?? 0;
    const bProps = config?.[b].properties?.length ?? 0;
    const aIsShorthand = SHORTHAND_PROPERTIES.includes(aProp) && aProps === 0;
    const bIsShorthand = SHORTHAND_PROPERTIES.includes(bProp) && bProps === 0;

    // True shorthands go first (border etc)
    if (aIsShorthand && !bIsShorthand) return SORT.A_BEFORE_B;
    if (bIsShorthand && !aIsShorthand) return SORT.B_BEFORE_A;

    if (aIsShorthand && bIsShorthand) {
      const aPriority = SHORTHAND_PROPERTIES.indexOf(aProp);
      const bPriority = SHORTHAND_PROPERTIES.indexOf(bProp);

      return compare(aPriority, bPriority);
    }

    // If its not a true short hand compare number of properties the prop is responsible for, longest first.
    return compare(aProps, bProps);
  });
