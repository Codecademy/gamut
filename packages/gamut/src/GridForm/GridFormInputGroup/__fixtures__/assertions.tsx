import { getComponent } from './renderers';

export const itHandlesAriaInvalid = (
  componentName: string,
  selector: string
): void => {
  describe('fields', () => {
    it('have the property aria-invalid', () => {
      hasAriaInvalidProperty(componentName, selector);
    });
    it('marks a field as aria-invalid when a field has an error', () => {
      isMarkedAriaInvalidIfErrorExists(componentName, selector);
    });
  });
};

const hasAriaInvalidProperty = (
  componentName: string,
  selector: string
): void => {
  const props = { validation: { required: true } };
  const component = getComponent(componentName, props);
  expect(component.find(selector).props()).toHaveProperty('aria-invalid');
  expect(component.find(selector).props()['aria-invalid']).toBeUndefined();
};

const isMarkedAriaInvalidIfErrorExists = (
  componentName: string,
  selector: string
): void => {
  const props = {
    validation: { required: 'Required' },
    error: true,
  };
  const component = getComponent(componentName, props);
  expect(component.find(selector).props()['aria-invalid']).toBeTruthy();
};
