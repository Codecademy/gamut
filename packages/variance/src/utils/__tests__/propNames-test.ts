import { testConfig } from '../__fixtures__/testConfig';
import { orderPropNames } from '../propNames';

describe('orderPropNames', () => {
  it('orders shorthand properties before longhand', () => {
    const result = orderPropNames({
      test2: {
        property: 'marginLeft',
      },
      test: {
        property: 'margin',
      },
    });

    expect(result).toEqual(['test', 'test2']);
  });

  it('orders higher precendence longhand properties before lower precendence', () => {
    const result = orderPropNames({
      test2: {
        property: 'border',
      },
      test: {
        property: 'borderTop',
      },
    });

    expect(result).toEqual(['test2', 'test']);
  });
  it('orders higher precendence longhand properties before many lower precedence props', () => {
    const result = orderPropNames({
      test2: {
        property: 'border',
      },
      test: {
        property: 'borderBottom',
      },
      test3: {
        property: 'borderTop',
      },
    });

    expect(result).toEqual(['test2', 'test3', 'test']);
  });
  it('orders non-shorthand properties by length of properties responsible', () => {
    const result = orderPropNames({
      marginX: {
        property: 'margin',
        properties: ['marginLeft', 'marginRight'],
      },
      marginY: {
        property: 'margin',
        properties: ['marginTop', 'marginBottom'],
      },
      margin: {
        property: 'margin',
        properties: ['marginLeft', 'marginRight', 'marginTop', 'marginBottom'],
      },
    });

    expect(result).toEqual(['margin', 'marginX', 'marginY']);
  });
  it('orders non-shorthand properties by length of properties responsible', () => {
    const result = orderPropNames({
      marginX: {
        property: 'margin',
        properties: ['marginLeft', 'marginRight'],
      },
      marginY: {
        property: 'margin',
        properties: ['marginTop', 'marginBottom'],
      },
      margin: {
        property: 'margin',
        properties: ['marginLeft', 'marginRight', 'marginTop', 'marginBottom'],
      },
      marginLeft: {
        property: 'marginLeft',
      },
    });

    expect(result).toEqual(['margin', 'marginX', 'marginY', 'marginLeft']);
  });
  it('orders all props', () => {
    const result = orderPropNames(testConfig);

    expect(result).toEqual([
      'border',
      'borderX',
      'borderY',
      'borderTop',
      'borderBottom',
      'borderLeft',
      'borderRight',
      'borderWidth',
      'borderWidthX',
      'borderWidthY',
      'borderStyle',
      'borderStyleX',
      'borderStyleY',
      'borderColor',
      'borderColorX',
      'borderColorY',
      'flex',
      'm',
      'mx',
      'my',
      'p',
      'px',
      'py',
      // end shorthands
      'textColor',
      'bg',
      'borderColorLeft',
      'borderColorRight',
      'borderColorTop',
      'borderColorBottom',
      'borderRadius',
      'borderRadiusLeft',
      'borderRadiusTop',
      'borderRadiusBottom',
      'borderRadiusRight',
      'flexBasis',
      'pt',
      'pb',
      'pr',
      'pl',
      'mt',
      'mb',
      'mr',
      'ml',
    ]);
  });
});
