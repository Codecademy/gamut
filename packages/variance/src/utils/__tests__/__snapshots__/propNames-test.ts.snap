// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orderPropNames orders all props 1`] = `
Array [
  "border",
  "borderX",
  "borderY",
  "borderTop",
  "borderBottom",
  "borderLeft",
  "borderRight",
  "borderWidth",
  "borderWidthX",
  "borderWidthY",
  "borderStyle",
  "borderStyleX",
  "borderStyleY",
  "borderColor",
  "borderColorX",
  "borderColorY",
  "borderRadius",
  "borderRadiusLeft",
  "borderRadiusTop",
  "borderRadiusBottom",
  "borderRadiusRight",
  "flex",
  "m",
  "mx",
  "my",
  "p",
  "px",
  "py",
  "inset",
  "size",
  "width",
  "top",
  "right",
  "pt",
  "pr",
  "pl",
  "pb",
  "minWidth",
  "minHeight",
  "maxWidth",
  "maxHeight",
  "mt",
  "mr",
  "ml",
  "mb",
  "left",
  "height",
  "flexBasis",
  "textColor",
  "bottom",
  "borderColorTop",
  "borderColorRight",
  "borderColorLeft",
  "borderColorBottom",
  "bg",
]
`;
